internal/xfn/container_linux.go

```diff
@@ -0,0 +1,161 @@
+	// Parse the input early, before we potentially pull and write the image.
+	y, err := yaml.Marshal(in)
+	if err != nil {
+		return nil, errors.Wrap(err, errInvalidInput)
+	}
+
+	/*
+		We want to create an overlayfs with the cached rootfs as the lower layer
+		and the bundle's rootfs as the upper layer, if possible. Kernel 5.11 and
+		later supports using overlayfs inside a user (and mount) namespace. The
+		best way to run code in a user namespace in Go is to execute a separate
+		binary; the unix.Unshare syscall affects only one OS thread, and the Go
+		scheduler might move the goroutine to another.
+
+		Therefore we execute a small shim - xfn spark - in a new user and mount
+		namespace. spark sets up the overlayfs if the Kernel supports it. It
+		then executes an OCI runtime which creates further namespaces in order
+		to actually execute the function.
+	*/
+	cmd := exec.CommandContext(ctx, os.Args[0], spark, "--cache-dir="+r.cache, r.image) //nolint:gosec // We're intentionally executing with variable input.
+	cmd.SysProcAttr = &syscall.SysProcAttr{
+		Cloneflags:  syscall.CLONE_NEWUSER | syscall.CLONE_NEWNS,
+		UidMappings: []syscall.SysProcIDMap{{ContainerID: 0, HostID: r.rootUID, Size: 1}},
+		GidMappings: []syscall.SysProcIDMap{{ContainerID: 0, HostID: r.rootGID, Size: 1}},
+	}
+
+	// When we have CAP_SETUID and CAP_SETGID (i.e. typically when root), we can
+	// map a range of UIDs (0 to 65,336) inside the user namespace to a range in
+	// its parent. We can also drop privileges (in the parent user namespace) by
+	// running spark as root in the user namespace.
+	if r.setuid {
+		cmd.SysProcAttr.UidMappings = []syscall.SysProcIDMap{{ContainerID: 0, HostID: r.rootUID, Size: UserNamespaceUIDs}}
+		cmd.SysProcAttr.GidMappings = []syscall.SysProcIDMap{{ContainerID: 0, HostID: r.rootGID, Size: UserNamespaceGIDs}}
+		cmd.SysProcAttr.GidMappingsEnableSetgroups = true
+
+		// UID and GID 0 here are relative to the new user namespace - i.e. they
+		// correspond to HostID in the parent. We're able to do this because
+		// Go's exec.Command will:
+		//
+		// 1. Call clone(2) to create a child process in a new user namespace.
+		// 2. In the child process, wait for /proc/self/uid_map to be written.
+		// 3. In the parent process, write the child's /proc/$pid/uid_map.
+		// 4. In the child process, call setuid(2) and setgid(2) per Credential.
+		// 5. In the child process, call execve(2) to execute spark.
+		//
+		// Per user_namespaces(7) the child process created by clone(2) starts
+		// out with a complete set of capabilities in the new user namespace
+		// until the call to execve(2) causes them to be recalculated. This
+		// includes the CAP_SETUID and CAP_SETGID necessary to become UID 0 in
+		// the child user namespace, effectively dropping privileges to UID
+		// 100000 in the parent user namespace.
+		//
+		// https://github.com/golang/go/blob/1b03568/src/syscall/exec_linux.go#L446
```

Comment:
Note to self: this giant comment block should use `/* */` style comments to match the other giant one above.